packages/Chem/src/chem-searches.ts

```diff
@@ -82,116 +86,151 @@ function _chemGetDiversities(limit: number, molStringsColumn: DG.Column, fingerp
 }
 
-function colInvalidated(col: DG.Column, createMols: boolean): Boolean {
+function colInvalidated(col: DG.Column, createMols: boolean, useSubstructLib?: boolean): Boolean {
   return (lastColumnInvalidated == invalidatedColumnKey(col) &&
     col.getTag(FING_COL_TAGS.invalidatedForVersion) == String(col.version) &&
-    (!createMols || col.getTag(FING_COL_TAGS.molsCreatedForVersion) == String(col.version)));
+    (!createMols || useSubstructLib ?
+      col.getTag(FING_COL_TAGS.substrLibCreatedForVersion) == String(col.version) :
+      col.getTag(FING_COL_TAGS.molsCreatedForVersion) == String(col.version)));
 }
 
-async function _invalidate(molCol: DG.Column, createMols: boolean) {
-  if (!colInvalidated(molCol, createMols)) {
+async function _invalidate(molCol: DG.Column, createMols: boolean, useSubstructLib?: boolean) {
+  if (!colInvalidated(molCol, createMols, useSubstructLib)) {
     if (createMols) {
-      await (await getRdKitService()).initMoleculesStructures(molCol.toList());
-      molCol.setTag(FING_COL_TAGS.molsCreatedForVersion, String(molCol.version + 2));
+      await (await getRdKitService()).initMoleculesStructures(molCol.toList(), useSubstructLib);
+      useSubstructLib ? molCol.setTag(FING_COL_TAGS.substrLibCreatedForVersion, String(molCol.version + 2)) :
+        molCol.setTag(FING_COL_TAGS.molsCreatedForVersion, String(molCol.version + 2));
     }
     lastColumnInvalidated = invalidatedColumnKey(molCol);
     molCol.setTag(FING_COL_TAGS.invalidatedForVersion, String(molCol.version + 1));
   }
 }
 
-function checkForFingerprintsColumn(col: DG.Column, fingerprintsType: Fingerprint): Uint8Array[] | null {
-  const colNameTag = '.' + fingerprintsType + '.Column';
-  const colVerTag = '.' + fingerprintsType + '.Version';
+
+function checkForSavedColumn(col: DG.Column, colTag: Fingerprint | string): DG.Column<any> | null{
+  const colNameTag = '.' + colTag + '.Column';
+  const colVerTag = '.' + colTag + '.Version';
 
   if (col.getTag(colNameTag) &&
       col.getTag(colVerTag) == String(col.version) &&
       col.dataFrame) {
-    const fingCol = col.dataFrame.columns.byName(col.getTag(colNameTag));
-    if (fingCol)
-      return fingCol.toList();
+    const savedCol = col.dataFrame.columns.byName(col.getTag(colNameTag));
+    return savedCol;
   }
   return null;
 }
 
-function saveFingerprintsToCol(col: DG.Column, fgs: Uint8Array[],
-  fingerprintsType: Fingerprint, createdMols: boolean): void {
+function saveColumns(col: DG.Column, data: ((Uint8Array | null)[] | (string | null)[])[],
+  tags: string[], colTypes: DG.COLUMN_TYPE[], createdMols: boolean): void {
+  
+  for (let i = 0; i < data.length; i++)
+    saveColumn(col, data[i], tags[i], colTypes[i]);
+
+  const colVersion = createdMols ? col.version + data.length + 2 : col.version + data.length + 1;
+  if (createdMols)
+    col.setTag(FING_COL_TAGS.molsCreatedForVersion, String(colVersion));
+  for (let i = 0; i < data.length; i++)
```

Comment:
Not sure I follow the logic. Why is the for loop goes to data.length, but we are settings tags[i]? Shouldn't it go to tags.length?

Reply:
Because data contains invisible columns which will be added to dataframe and for each added column we will add a tag containing column version. And the final column version depends on the number of added tags. 
For instance we are adding both pattern fp and canonical smiles. So we are adding 2 tags:
'.Pattern.version' and '.canonicalSmiles.Version'. Anyway, I'm now looking at all this logic with tags and I think this should be simplified. Will look into it once more